crates/bevy_pbr/src/meshlet/gpu_scene.rs

```diff
@@ -0,0 +1,763 @@
+                    ShaderStages::FRAGMENT,
+                    (
+                        storage_buffer_read_only_sized(false, None),
+                        storage_buffer_read_only_sized(false, None),
+                        storage_buffer_read_only_sized(false, None),
+                        storage_buffer_read_only_sized(false, None),
+                        storage_buffer_read_only_sized(false, None),
+                        storage_buffer_read_only_sized(false, None),
+                        storage_buffer_read_only_sized(false, None),
+                        texture_2d(TextureSampleType::Uint),
+                    ),
+                ),
+            ),
+            depth_pyramid_sampler: render_device.create_sampler(&SamplerDescriptor {
+                label: Some("meshlet_depth_pyramid_sampler"),
+                ..default()
+            }),
+        }
+    }
+}
+
+impl MeshletGpuScene {
+    fn reset(&mut self) {
+        // TODO: Shrink capacity if saturation is low
+        self.scene_meshlet_count = 0;
+        self.scene_index_count = 0;
+        self.next_material_id = 0;
+        self.material_id_lookup.clear();
+        self.material_ids_present_in_scene.clear();
+        self.instances.clear();
+        self.instance_uniforms.get_mut().clear();
+        self.instance_material_ids.get_mut().clear();
+        self.thread_instance_ids.get_mut().clear();
+        self.thread_meshlet_ids.get_mut().clear();
+        self.previous_thread_ids.get_mut().clear();
+        self.previous_thread_id_starts
+            .values_mut()
+            .for_each(|(_, active)| *active = false);
+    }
+
+    fn queue_meshlet_mesh_upload(
+        &mut self,
+        instance: Entity,
+        handle: &Handle<MeshletMesh>,
+        assets: &mut Assets<MeshletMesh>,
+        instance_index: u32,
+    ) {
+        let queue_meshlet_mesh = |asset_id: &AssetId<MeshletMesh>| {
+            let meshlet_mesh = assets.remove_untracked(*asset_id).expect(
+                "MeshletMesh asset was already unloaded but is not registered with MeshletGpuScene",
+            );
+
+            let vertex_data_slice = self
+                .vertex_data
+                .queue_write(Arc::clone(&meshlet_mesh.vertex_data), ());
+            let vertex_ids_slice = self.vertex_ids.queue_write(
+                Arc::clone(&meshlet_mesh.vertex_ids),
+                vertex_data_slice.start,
+            );
+            let indices_slice = self
+                .indices
+                .queue_write(Arc::clone(&meshlet_mesh.indices), ());
+            let meshlets_slice = self.meshlets.queue_write(
+                Arc::clone(&meshlet_mesh.meshlets),
+                (vertex_ids_slice.start, indices_slice.start),
+            );
+            self.meshlet_bounding_spheres
+                .queue_write(Arc::clone(&meshlet_mesh.meshlet_bounding_spheres), ());
+
+            (
+                (meshlets_slice.start as u32 / 12)..(meshlets_slice.end as u32 / 12),
```

Comment:
why is this 12? is it 4 byte alignment * 3 indices per triangle?

Reply:
Can't remember off the top of my head, but I'll figure it out and document it :sweat_smile:. I didn't write any documentation yet, it's sorely needed. This PR is probably very hard to review otherwise.